2289 add use existingtask tests

zapper59 opened the pull request.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "2289-Add-useExistingtask-tests" git@github.com:zapper59/swarmkit.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354509552
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

zapper59 commented:
Fixes https://github.com/docker/swarmkit/issues/2289

anshulpundir commented on manager/orchestrator/update/updater_test.go:
```diff
@@ -592,3 +592,195 @@ func TestUpdaterOrder(t *testing.T) {
 		}
 	}
 }
+
```
Add a high level comment on the test-case(s) covered.

flippedAben replied:
Yes, we can do that.

zapper59 commented:
Would you happen to know how I can get rid of/ do I need to get rid of the dco/no label?  Is it there because the code coverage report is confused?

anshulpundir commented:
https://github.com/docker/swarmkit/pull/2466#issuecomment-348749832 @zapper59

zapper59 commented:
@anshulpundir Thanks!  We were able to figure out how to figure how to remove dco/no.  Also it appears that codecov is reporting decreased code coverage from our pull request because the last merged PR had failing tests and did not include several packages.  For reference our tests appear to increase coverage of github.com/docker/swarmkit/manager/orchestrator/update from 58.9% to 68.1%.

anshulpundir commented on manager/orchestrator/update/updater_test.go:
```diff
@@ -592,3 +592,197 @@ func TestUpdaterOrder(t *testing.T) {
+		},
+	}
+
+	// 1: First test calls Run with one slot of two running tasks and
```
nit: First test calls Run with one slot of two running tasks => First test creates two tasks for the same slot and calls updater.Run() and expects one task to be running and one to be shutdown.

anshulpundir commented on manager/orchestrator/update/updater_test.go:
```diff
@@ -592,3 +592,197 @@ func TestUpdaterOrder(t *testing.T) {
+		}
+	}
+
+	//  3: We create another slot of 2 tasks that desire to be ready and expect
```
nit: We create another slot of 2 tasks that desire to be ready =>  We create another slot with 2 tasks with desired state READY

anshulpundir commented on manager/orchestrator/update/updater_test.go:
```diff
@@ -592,3 +592,197 @@ func TestUpdaterOrder(t *testing.T) {
+	//  4: We run with two ready tasks where the second is dirty
+	//  and expect the first to run and the second to shutdown
+	err = s.Update(func(tx store.Tx) error {
+		firstTask.DesiredState = api.TaskStateShutdown
```
this seems similar to the previous test case. Did I miss something ?

zapper59 commented:
@anshulpundir I've updated the comments describing test cases.  Also I tweeked the 4th case to something that seems more relevant.

dperny commented on manager/orchestrator/update/updater_test.go:
```diff
@@ -592,3 +592,199 @@ func TestUpdaterOrder(t *testing.T) {
+		} else if task.ID == secondTask.ID {
+			assert.Equal(t, task.DesiredState, api.TaskStateShutdown)
+		} else {
+			assert.True(t, false)
```
lol, don't use assert here, just fall back onto the usual go testing functions
```go
t.Errorf("Got unexpected task %v in the task listing, should only have gotten %v and %v",
    firstTask.ID, secondTask.ID,
)
```

dperny commented:
This uses `assert` everywhere, which I think is the pattern used in the other tests. but assert allows a test to continue running after an assertion has failed. In places where it wouldn't make sense to continue the test, you can swap the `assert` for `require`; this is mostly the case for things like setup, errors during which basically guarantee the rest of the test won't run correctly. 

The major caveat is that you cannot use `require` inside of a goroutine! Only inside of the main test routine! The `require` will abort the test by calling runtime.Goexit, which runs defers, but it _only_ terminates the current goroutine.

The second thing I'd like is some messages on the asserts and requires. They all can take a format string as extra arguments, just like `t.Errorf`, and you can use that to explain in more detail what's gone wrong with the test, somewhat like this:

```go
assert.True(t, (tasks[0].DesiredState == api.TaskStateRunning &&
	tasks[1].DesiredState == api.TaskStateShutdown) ||
	(tasks[1].DesiredState == api.TaskStateRunning &&
		tasks[0].DesiredState == api.TaskStateShutdown),
	"Expected on task to be desired RUNNING and the other to be desired SHUTDOWN, but they were in %v and %v",
	tasks[0].DesiredState, tasks[1].DesiredState,
)
```

If the test flakes later, this makes it way easier to quickly debug. You don't have to include this in EVERY statement, just the ones where it's not easy to ascertain what the failure condition was.

zapper59 commented:
@dperny How do our latest test updates look?  I think we've changed all of the relevant asserts to requires, and added a few formatted strings to error messages.
